src/exodus/test/sigmawallet_tests.cpp

```diff
@@ -0,0 +1,328 @@
+
+    BOOST_CHECK_EQUAL(false, sigmaWallet.TryRecoverMint(
+        id,
+        SigmaMintChainState(1000, 0, 1000)
+    ));
+
+    // verify state after, mint wallet should not be changed
+    std::vector<MintPoolEntry> mintPoolAfter;
+    sigmaWallet.GetMintPoolEntry(std::back_inserter(mintPoolAfter));
+
+    BOOST_CHECK(mintPool == mintPoolAfter);
+
+    BOOST_CHECK_EQUAL(true, sigmaWallet.HasMint(id));
+}
+
+BOOST_AUTO_TEST_CASE(listmints_empty_wallet)
+{
+    std::vector<SigmaMint> mints;
+    sigmaWallet.ListMints(std::back_inserter(mints));
+    BOOST_CHECK_EQUAL(0, mints.size());
+}
+
+BOOST_AUTO_TEST_CASE(listmints_non_empty_wallet)
+{
+    // generate 3 coins which is
+    // 1. unconfirmed
+    // 2. confirmed and unspend
+    // 3. spend
+    PropertyId prop = 10;
+    SigmaDenomination denom = 0;
+
+    auto unconfirmed = sigmaWallet.GenerateMint(prop, denom);
+    auto unspend = sigmaWallet.GenerateMint(prop, denom);
+    sigmaWallet.UpdateMintChainstate(
+        SigmaMintId(prop, denom, SigmaPublicKey(unspend.second, DefaultSigmaParams)),
+        SigmaMintChainState(100, 0, 1000));
+
+    auto spend = sigmaWallet.GenerateMint(prop, denom);
+    sigmaWallet.UpdateMintChainstate(
+        SigmaMintId(prop, denom, SigmaPublicKey(spend.second, DefaultSigmaParams)),
+        SigmaMintChainState(100, 0, 1001));
+
+    sigmaWallet.UpdateMintSpendTx(
+        SigmaMintId(prop, denom, SigmaPublicKey(spend.second, DefaultSigmaParams)),
+        uint256S("1"));
+
+    // prepare testing function
+    auto sigmaMintComparer = [](SigmaMint const &a, SigmaMint const &b) -> bool {
+        return a.property == b.property &&
+            a.denomination == b.denomination &&
+            a.seedId == b.seedId;
+    };
+
+
+
+    std::vector<SigmaMint> mints;
+    sigmaWallet.ListMints(std::back_inserter(mints));
+    BOOST_CHECK_EQUAL(
+        true,
+        std::is_permutation(
+            mints.begin(), mints.end(),
+            std::vector<SigmaMint>{
+                unconfirmed.first, unspend.first, spend.first
+            }.begin(),
+            sigmaMintComparer
+        )
+    );
+}
+
+BOOST_AUTO_TEST_CASE(push_out_wallet_mint_back)
+{
+    SigmaPrivateKey privKey;
+    privKey.Generate();
+    SigmaPublicKey pubKey(privKey, DefaultSigmaParams);
+
+    SigmaMintId id(10, 0, pubKey);
+
+    BOOST_CHECK_EXCEPTION(
+        sigmaWallet.DeleteUnconfirmedMint(id),
+        std::runtime_error,
+        [](std::runtime_error const &e) -> bool{
+            return std::string("no mint data in wallet") == e.what();
+        }
+    );
+}
+
+BOOST_AUTO_TEST_CASE(push_mint_back)
+{
+    std::vector<MintPoolEntry> mintPool;
+    sigmaWallet.GetMintPoolEntry(std::back_inserter(mintPool));
+
+    SigmaMint mint;
+    SigmaPrivateKey privKey;
+    std::tie(mint, privKey) = sigmaWallet.GenerateMint(1, 0);
+
+    SigmaMintId id(1, 0, SigmaPublicKey(privKey, DefaultSigmaParams));
+
+    sigmaWallet.DeleteUnconfirmedMint(id);
```

Comment:
We should check if mint exists before we delete and not exists after we delete too.